adal/src/main/java/com/microsoft/aad/adal/AcquireTokenSilentHandler.java

```diff
@@ -164,7 +164,7 @@ AuthenticationResult acquireTokenWithRefreshToken(final String refreshToken)
 
             throw new AuthenticationException(
                     ADALError.AUTH_FAILED_NO_TOKEN, ExceptionExtensions.getExceptionMessage(exc),
-                    new AuthenticationException(ADALError.SERVER_ERROR, exc.getMessage()));
+                    new AuthenticationException(ADALError.SERVER_ERROR, exc.getMessage(), exc));
```

Comment:
Looking at this code path, there might be something misleading going on here... If this catches a lower Exception thrown for `4XX`-level HTTP status codes (say, a `429`), these can't accurately be called instances of `SERVER_ERROR` since `4XX` error codes are reserved for client errors (Bad Request, Unauthorized, Forbidden, Not Found, etc) -- in some instances `INVALID_REQUEST` may be _more_ appropriate but this also seems like the wrong place to start conditionally checking thrown Exceptions...

Since this code path _may_ be rethrowing an `AuthenticationException`, can we check if there's an existing `ADALError` that may be populated on the Exception before rethrowing, rather than assume it's a `SERVER_ERROR`?

Reply:
Good point, this part confused me as well. Currently adal we don't have specific ADALError enum constant for every http status code or status code group like 4XX, we treat all status code (except for 200, 400, 401) as ADALError.SERVER_ERROR ("Server returned an error"). The developer can get the specific error in String by calling AuthenticationException.getResponseBody().get(AuthenticationConstants.OAuth2.ERROR).

The HTTP web response error is a String type, not an ADALError. We will need the full list of error code if we want to subdivide the error code here. Please let me know you thoughts. @iambmelt @addev-ashish I'm open mind on how we label the http status code in AuthenticationException.

Reply:
For this case, I might suggest something _roughly_ equivalent to the following:

```java
new AuthenticationException(
        null == exc.getCode() ? ADALError.SERVER_ERROR : exc.getCode(),
        exc.getMessage(),
        exc
    ));
```

Clearly this doesn't solve the _larger issue_ of possibly returning the wrong `ADALError` instance, but I can live with this for now.

I'm fine to make an issue to review our thrown errors and revisit as part of another cleanup/refactor effort.

/cc @addev-ashish @nazukj

Reply:
Sorry, @iambmelt I don't understand why we need to do the check here. The AuthenticationException error code should take the outermost layer's value. In this case, in Line 166, the error code is AUTH_FAILED_NO_TOKEN. We are not replacing the error code with ADALError.SERVER_ERROR. I'm confused why we need to add this checking.
```java
new AuthenticationException(
        null == exc.getCode() ? ADALError.SERVER_ERROR : exc.getCode(),
        exc.getMessage(),
        exc
    ));
```
The ADALError.SERVER_ERROR is a general error code for all http error response, not for specific status code. I totally agree the name itself is misleading and we should add more ADALError to represent the status code better. But I'm not sure if this is the purpose of issue #1046 

And if the developer want to get the http web response error code, for example, if we get the http response body like this
{"error":"interaction_required",
"error_description":"AADSTS50131: Windows device is not in required device state: domain_joined\r\nTrace ID: 41f28fdd-8c37-4f3e-8e51-2eeea28e0700\r\nCorrelation ID: 058ff775-5dfe-4e0a-901f-2440103cfb59\r\nTimestamp: 2017-10-11 00:20:23Z",
"error_codes":[50131],
"timestamp":"2017-10-11 00:20:23Z",
"trace_id":"41f28fdd-8c37-4f3e-8e51-2eeea28e0700",
"correlation_id":"058ff775-5dfe-4e0a-901f-2440103cfb59"}
they can call AuthenticationException.getResponseBody().get(AuthenticationConstants.OAuth2.ERROR) and get "interaction_required".

@addev-ashish Please help me to confirm if there is anything I misunderstand. Thanks. :)

Reply:
Thanks for clarifying, @heidijinxujia -- I'll say :shipit:

Reply:
I filed the issue we discussed here #1068 @iambmelt